client/src/main/java/io/pravega/client/netty/impl/RawClient.java

```diff
@@ -0,0 +1,141 @@
+import io.pravega.common.auth.AuthenticationException;
+import io.pravega.shared.protocol.netty.ConnectionFailedException;
+import io.pravega.shared.protocol.netty.FailingReplyProcessor;
+import io.pravega.shared.protocol.netty.PravegaNodeUri;
+import io.pravega.shared.protocol.netty.Reply;
+import io.pravega.shared.protocol.netty.Request;
+import io.pravega.shared.protocol.netty.WireCommand;
+import io.pravega.shared.protocol.netty.WireCommands;
+import io.pravega.shared.protocol.netty.WireCommands.Hello;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.annotation.concurrent.GuardedBy;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class RawClient implements AutoCloseable {
+
+    private final CompletableFuture<ClientConnection> connection;
+    
+    private final Object lock = new Object();
+    @GuardedBy("lock")
+    private final Map<Long, CompletableFuture<Reply>> requests = new HashMap<>();
+    private final ResponseProcessor responseProcessor = new ResponseProcessor();
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+    private final class ResponseProcessor extends FailingReplyProcessor {
+        
+        @Override
+        public void process(Reply reply) {
+            if (reply instanceof Hello) {
+                Hello hello = (Hello) reply;
+                log.info("Received hello: {}", hello);
+                if (hello.getLowVersion() > WireCommands.WIRE_VERSION || hello.getHighVersion() < WireCommands.OLDEST_COMPATIBLE_VERSION) {
+                    closeConnection(new IllegalStateException("Incompatible wire protocol versions " + hello));
+                }
+            } else {
+                log.debug("Received reply {}", reply);
+                reply(reply);
+            }
+        }
+        
+        @Override
+        public void connectionDropped() {
+            closeConnection(new ConnectionFailedException());
+        }
+
+        @Override
+        public void processingFailure(Exception error) {
+            log.warn("Processing failure: ", error);
+            closeConnection(error);
+        }
+        
+        @Override
+        public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenCheckFailed) {
+            log.warn("Auth token failure: ", authTokenCheckFailed);
+            closeConnection(new AuthenticationException(authTokenCheckFailed.toString()));
+        }
+    }
+    
+    public RawClient(Controller controller, ConnectionFactory connectionFactory, Segment segmentId) {
+        connection = controller.getEndpointForSegment(segmentId.getScopedName())
+                               .thenCompose((PravegaNodeUri uri) -> connectionFactory.establishConnection(uri, responseProcessor));
+        connection.exceptionally(e -> {
```

Comment:
1. No need to re-throw the exception as nobody is listening to this callback.
2. For this particular case, I suggest using `Futures.exceptionListener`. Makes your code more compact.

Reply:
Done.